### PR TITLE
insert vertex & edge if not exists

### PR DESCRIPTION
**Pull request by nevermore3**
depend on https://github.com/vesoft-inc/nebula-common/pull/473

### CODE REVIEW

File: src/storage/mutate/AddVerticesProcessor.cpp [hunk 1]
```diff
@@ -160,25 +179,23 @@ void AddVerticesProcessor::doProcessWithIndex(const cpp2::AddVerticesRequest& re
                 }
 
                 auto key = NebulaKeyUtils::vertexKey(spaceVidLen_, partId, vid, tagId);
+                if (ifNotExists_ && !visited.emplace(key).second) {
```
Comment (critical27) [inline src/storage/mutate/AddVerticesProcessor.cpp:182]: What is the expected order when graph insert same (vId + tagId) with different value?

File: src/storage/mutate/AddVerticesProcessor.cpp [hunk 2]
```diff
@@ -92,6 +94,20 @@ void AddVerticesProcessor::doProcess(const cpp2::AddVerticesRequest& req) {
                 }
 
                 auto key = NebulaKeyUtils::vertexKey(spaceVidLen_, partId, vid, tagId);
+                if (ifNotExists_) {
+                    if (!visited.emplace(key).second) {
+                        continue;
+                    }
+                    auto obsIdx = findOldValue(partId, vid, tagId);
```
Comment (critical27) [inline src/storage/mutate/AddVerticesProcessor.cpp:101]: Maybe we should emplace result into vertexCache_ in `findOldValue`, at least add a todo

**Review by critical27 — COMMENTED** [review]
We don't have a good read-modify-write api (asyncAtomicOp could do it, but hard to use) in storage, what is the decision about this `if_not_exists`?

**Comment by nevermore3** [comment]
> We don't have a good read-modify-write api (asyncAtomicOp could do it, but not so friendly) in storage, what is the decision about this `if_not_exists`?

some community users need this feature ,please refer to https://jira.nebula-graph.io/browse/NG-420

File: src/storage/mutate/AddVerticesProcessor.cpp (hunk 1 above)
Comment (nevermore3) [reply src/storage/mutate/AddVerticesProcessor.cpp:182]: in order of input, if encounter the same(vid + tagid) the do nothing when if_not_exists

File: src/storage/mutate/AddVerticesProcessor.cpp (hunk 2 above)
Comment (nevermore3) [reply src/storage/mutate/AddVerticesProcessor.cpp:101]: make sense

**Comment by bright-starry-sky** [comment]
Good job, Three questions： 
1，insert vertex t1(c1) values 1:(1), 1(2), 1(3)... , what is expected to be inserted? 
2, Do we need to consider the index?
3, How about performance?

**Comment by nevermore3** [comment]
> Good job, Three questions：
> 1，insert vertex t1(c1) values 1:(1), 1(2), 1(3)... , what is expected to be inserted?
> 2, Do we need to consider the index?
> 3, How about performance?

1 .  only 1:(1) should be inserted
2. Yes, But the processing of the index is after this logic, so it does not affect the index
3.  In this scenario, performance will inevitably decrease, especially when there is no index.

File: src/storage/mutate/AddVerticesProcessor.cpp (hunk 1 above)
Comment (critical27) [reply src/storage/mutate/AddVerticesProcessor.cpp:182]: So when `ifNotExists_` is true, first one will be inserted, when `ifNotExists_` is false, the last one will be insert. Am I right?

File: src/storage/mutate/AddVerticesProcessor.cpp (hunk 2 above)
Comment (nevermore3) [reply src/storage/mutate/AddVerticesProcessor.cpp:101]: Do you mean vetexCache_->evict(vid,tagId) if (vid,tagId) already exists in kvstroe In the findOldValue function?, i'm a little confused

File: src/storage/mutate/AddVerticesProcessor.cpp (hunk 1 above)
Comment (nevermore3) [reply src/storage/mutate/AddVerticesProcessor.cpp:182]: right

**Review by Aiee — DISMISSED** [review]

File: src/storage/mutate/AddVerticesProcessor.cpp (hunk 2 above)
Comment (critical27) [reply src/storage/mutate/AddVerticesProcessor.cpp:101]: Sorry for late reply, I mean we could insert to cache if we find an old value. The logic could be handled in another PR,  the logic is weird indeed.

Comment (nevermore3) [reply src/storage/mutate/AddVerticesProcessor.cpp:101]: It's okay, I understand what you mean

**Review by critical27 — APPROVED** [review]
LGTM

**Review by CPWstatic — APPROVED** [review]

**Review by critical27 — APPROVED** [review]

**Review by yixinglu — APPROVED** [review]

**Review by critical27 — APPROVED** [review]

**Review by liuyu85cn — APPROVED** [review]

**Pull request merged**